File: frontend/lib/circuitBreaker.ts
/**
 * Circuit Breaker and Risk Management System
 *
 * Migrated from Polymarket-Kalshi-Arbitrage-bot circuit_breaker.rs
 * Implements production-grade risk management with configurable limits and automatic shutoffs
 *
 * Key Features:
 * - Position limits per market and total
 * - Daily loss limits with automatic trading halt
 * - Per-trade loss limits
 * - Consecutive error tracking
 * - Gas cost buffers
 * - Automatic circuit breaker trip and recovery
 */

// ============================================================================
// TYPES
// ============================================================================

export enum CircuitBreakerState {
  CLOSED = 'CLOSED', // Trading enabled (normal operation)
  OPEN = 'OPEN', // Trading disabled (circuit tripped)
  HALF_OPEN = 'HALF_OPEN', // Testing if conditions improved
}

export interface CircuitBreakerConfig {
  // Position limits
  max_position_per_market: number; // Max contracts per market (default: 50,000)
  max_total_position: number; // Max total contracts (default: 100,000)

  // Loss limits
  max_daily_loss_usd: number; // Daily loss limit (default: $500)
  max_loss_per_trade_usd: number; // Per-trade loss limit (default: $5)

  // Error tracking
  max_consecutive_errors: number; // Max consecutive failures (default: 5)
  error_cooldown_ms: number; // Cooldown after trip (default: 60,000ms = 1min)

  // Gas costs
  gas_buffer_cents: number; // Gas buffer per order (default: 3¢)

  // Position sizing
  liquidity_factor: number; // Fraction of liquidity to use (default: 0.5)
}

export const DEFAULT_CIRCUIT_BREAKER_CONFIG: CircuitBreakerConfig = {
  max_position_per_market: 50000,
  max_total_position: 100000,
  max_daily_loss_usd: 500,
  max_loss_per_trade_usd: 5,
  max_consecutive_errors: 5,
  error_cooldown_ms: 60000, // 1 minute
  gas_buffer_cents: 3,
  liquidity_factor: 0.5,
};

export interface Position {
  market_id: string;
  question: string;
  yes_quantity: number; // Number of YES contracts
  no_quantity: number; // Number of NO contracts
  avg_yes_price: number; // Average entry price for YES
  avg_no_price: number; // Average entry price for NO
  current_value_usd: number; // Current market value
  unrealized_pnl_usd: number; // Unrealized P&L
  timestamp: number; // Last update
}

export interface TradeResult {
  success: boolean;
  error_message?: string;
  execution_time_ms?: number;
  actual_slippage_cents?: number;
  gas_cost_usd?: number;
  position?: Position;
}

export interface DailyMetrics {
  date: string; // YYYY-MM-DD
  total_trades: number;
  successful_trades: number;
  failed_trades: number;
  total_pnl_usd: number;
  max_drawdown_usd: number;
  consecutive_errors: number;
  total_gas_spent_usd: number;
}

// ============================================================================
// CIRCUIT BREAKER CLASS
// ============================================================================

export class CircuitBreaker {
  private config: CircuitBreakerConfig;
  private state: CircuitBreakerState;
  private positions: Map<string, Position>;
  private daily_metrics: DailyMetrics;
  private trip_time: number | null;
  private error_count: number;
  private last_reset_time: number;

  constructor(config: CircuitBreakerConfig = DEFAULT_CIRCUIT_BREAKER_CONFIG) {
    this.config = config;
    this.state = CircuitBreakerState.CLOSED;
    this.positions = new Map();
    this.trip_time = null;
    this.error_count = 0;
    this.last_reset_time = Date.now();

    // Initialize daily metrics
    this.daily_metrics = this.initializeDailyMetrics();
  }

  // ========================================================================
  // STATE MANAGEMENT
  // ========================================================================

  /**
   * Get current circuit breaker state
   */
  getState(): CircuitBreakerState {
    // Auto-transition from HALF_OPEN to CLOSED if conditions improved
    if (
      this.state === CircuitBreakerState.HALF_OPEN &&
      this.canReset()
    ) {
      this.state = CircuitBreakerState.CLOSED;
      this.error_count = 0;
      this.trip_time = null;
    }

    // Auto-transition from OPEN to HALF_OPEN after cooldown
    if (
      this.state === CircuitBreakerState.OPEN &&
      this.trip_time &&
      Date.now() - this.trip_time > this.config.error_cooldown_ms
    ) {
      this.state = CircuitBreakerState.HALF_OPEN;
      this.error_count = Math.floor(this.error_count / 2); // Reduce error count
    }

    return this.state;
  }

  /**
   * Check if trading is allowed
   */
  canTrade(): boolean {
    const state = this.getState();
    return state === CircuitBreakerState.CLOSED || state === CircuitBreakerState.HALF_OPEN;
  }

  /**
   * Check if circuit breaker can be reset
   */
  private canReset(): boolean {
    return (
      this.error_count === 0 &&
      this.daily_metrics.consecutive_errors < this.config.max_consecutive_errors &&
      this.daily_metrics.total_pnl_usd > -this.config.max_daily_loss_usd
    );
  }

  /**
   * Trip the circuit breaker (halt trading)
   */
  private trip(reason: string): void {
    this.state = CircuitBreakerState.OPEN;
    this.trip_time = Date.now();
    console.error(`[CIRCUIT BREAKER] TRIPPED: ${reason}`);
  }

  // ========================================================================
  // TRADE VALIDATION
  // ========================================================================

  /**
   * Validate if a trade can be executed
   *
   * Checks:
   * 1. Circuit breaker state
   * 2. Position limits (per market and total)
   * 3. Daily loss limit
   * 4. Per-trade loss limit
   *
   * @returns {canExecute: boolean, reason?: string}
   */
  validateTrade(
    market_id: string,
    trade_size_usd: number,
    estimated_loss_usd: number
  ): { canExecute: boolean; reason?: string } {
    // Check circuit breaker state
    if (!this.canTrade()) {
      return {
        canExecute: false,
        reason: `Circuit breaker is ${this.state}`,
      };
    }

    // Check daily loss limit
    const projected_loss = this.daily_metrics.total_pnl_usd - estimated_loss_usd;
    if (projected_loss < -this.config.max_daily_loss_usd) {
      this.trip(`Daily loss limit exceeded: $${projected_loss.toFixed(2)}`);
      return {
        canExecute: false,
        reason: `Daily loss limit would be exceeded`,
      };
    }

    // Check per-trade loss limit
    if (estimated_loss_usd > this.config.max_loss_per_trade_usd) {
      return {
        canExecute: false,
        reason: `Per-trade loss limit exceeded: $${estimated_loss_usd.toFixed(2)}`,
      };
    }

    // Check position limits
    const current_position = this.positions.get(market_id);
    const current_position_size = current_position
      ? current_position.yes_quantity + current_position.no_quantity
      : 0;

    const new_position_size = current_position_size + trade_size_usd;

    if (new_position_size > this.config.max_position_per_market) {
      return {
        canExecute: false,
        reason: `Position limit for market would be exceeded: ${new_position_size} > ${this.config.max_position_per_market}`,
      };
    }

    // Check total position limit
    const total_position = this.calculateTotalPosition();
    if (total_position + trade_size_usd > this.config.max_total_position) {
      return {
        canExecute: false,
        reason: `Total position limit would be exceeded`,
      };
    }

    return { canExecute: true };
  }

  // ========================================================================
  // POSITION TRACKING
  // ========================================================================

  /**
   * Update position after trade execution
   */
  updatePosition(market_id: string, result: TradeResult): void {
    if (!result.success) {
      this.handleError(result.error_message || 'Unknown error');
      return;
    }

    // Reset error count on success
    this.error_count = 0;
    this.daily_metrics.consecutive_errors = 0;

    // Update position (simplified - in real system would track actual fills)
    let position = this.positions.get(market_id);
    if (!position && result.position) {
      position = result.position;
      this.positions.set(market_id, position);
    }
  }

  /**
   * Get position for a market
   */
  getPosition(market_id: string): Position | undefined {
    return this.positions.get(market_id);
  }

  /**
   * Get all positions
   */
  getAllPositions(): Position[] {
    return Array.from(this.positions.values());
  }

  /**
   * Calculate total position size across all markets
   */
  private calculateTotalPosition(): number {
    let total = 0;
    for (const position of this.positions.values()) {
      total += position.yes_quantity + position.no_quantity;
    }
    return total;
  }

  /**
   * Calculate unrealized P&L for all positions
   */
  calculateUnrealizedPnL(): number {
    let total_pnl = 0;
    for (const position of this.positions.values()) {
      total_pnl += position.unrealized_pnl_usd;
    }
    return total_pnl;
  }

  // ========================================================================
  // ERROR HANDLING
  // ========================================================================

  /**
   * Handle trade execution error
   */
  private handleError(error_message: string): void {
    this.error_count++;
    this.daily_metrics.consecutive_errors++;
    this.daily_metrics.failed_trades++;

    console.error(`[CIRCUIT BREAKER] Error (${this.error_count}/${this.config.max_consecutive_errors}): ${error_message}`);

    // Trip circuit breaker if too many consecutive errors
    if (this.error_count >= this.config.max_consecutive_errors) {
      this.trip(`Too many consecutive errors: ${this.error_count}`);
    }
  }

  /**
   * Record successful trade
   */
  recordSuccess(trade_result: TradeResult): void {
    this.daily_metrics.successful_trades++;
    this.daily_metrics.total_trades++;

    // Update P&L
    if (trade_result.position) {
      this.daily_metrics.total_pnl_usd += trade_result.position.unrealized_pnl_usd;
    }

    // Track gas costs
    if (trade_result.gas_cost_usd) {
      this.daily_metrics.total_gas_spent_usd += trade_result.gas_cost_usd;
    }

    // Reset error count on success
    this.error_count = 0;
    this.daily_metrics.consecutive_errors = 0;
  }

  // ========================================================================
  // METRICS AND REPORTING
  // ========================================================================

  /**
   * Get daily metrics
   */
  getDailyMetrics(): DailyMetrics {
    // Reset if new day
    const today = new Date().toISOString().split('T')[0];
    if (this.daily_metrics.date !== today) {
      this.daily_metrics = this.initializeDailyMetrics();
    }

    return { ...this.daily_metrics };
  }

  /**
   * Initialize daily metrics
   */
  private initializeDailyMetrics(): DailyMetrics {
    const today = new Date().toISOString().split('T')[0];
    return {
      date: today,
      total_trades: 0,
      successful_trades: 0,
      failed_trades: 0,
      total_pnl_usd: 0,
      max_drawdown_usd: 0,
      consecutive_errors: 0,
      total_gas_spent_usd: 0,
    };
  }

  /**
   * Get circuit breaker status
   */
  getStatus(): {
    state: CircuitBreakerState;
    can_trade: boolean;
    error_count: number;
    consecutive_errors: number;
    daily_pnl_usd: number;
    daily_loss_remaining_usd: number;
    total_positions: number;
    trip_time: number | null;
  } {
    const metrics = this.getDailyMetrics();
    const state = this.getState();

    return {
      state,
      can_trade: this.canTrade(),
      error_count: this.error_count,
      consecutive_errors: metrics.consecutive_errors,
      daily_pnl_usd: metrics.total_pnl_usd,
      daily_loss_remaining_usd: this.config.max_daily_loss_usd - Math.abs(metrics.total_pnl_usd),
      total_positions: this.positions.size,
      trip_time: this.trip_time,
    };
  }

  /**
   * Get detailed diagnostics
   */
  getDiagnostics(): {
    config: CircuitBreakerConfig;
    state: CircuitBreakerState;
    positions: Position[];
    daily_metrics: DailyMetrics;
    status: ReturnType<CircuitBreaker['getStatus']>;
  } {
    return {
      config: { ...this.config },
      state: this.getState(),
      positions: this.getAllPositions(),
      daily_metrics: this.getDailyMetrics(),
      status: this.getStatus(),
    };
  }

  // ========================================================================
  // MANUAL CONTROLS
  // ========================================================================

  /**
   * Manually reset circuit breaker (use with caution)
   */
  manualReset(): void {
    console.warn('[CIRCUIT BREAKER] Manual reset requested');
    this.state = CircuitBreakerState.CLOSED;
    this.error_count = 0;
    this.trip_time = null;
    this.daily_metrics.consecutive_errors = 0;
  }

  /**
   * Manually trip circuit breaker (emergency stop)
   */
  manualTrip(reason: string = 'Manual trip'): void {
    console.warn('[CIRCUIT BREAKER] Manual trip requested');
    this.trip(reason);
  }

  /**
   * Update configuration
   */
  updateConfig(updates: Partial<CircuitBreakerConfig>): void {
    this.config = { ...this.config, ...updates };
    console.log('[CIRCUIT BREAKER] Configuration updated:', this.config);
  }
}

// ============================================================================
// UTILITY FUNCTIONS
// ============================================================================

/**
 * Create a circuit breaker instance with default configuration
 */
export function createCircuitBreaker(
  config?: Partial<CircuitBreakerConfig>
): CircuitBreaker {
  const fullConfig = config
    ? { ...DEFAULT_CIRCUIT_BREAKER_CONFIG, ...config }
    : DEFAULT_CIRCUIT_BREAKER_CONFIG;
  return new CircuitBreaker(fullConfig);
}

/**
 * Validate circuit breaker configuration
 */
export function validateConfig(config: CircuitBreakerConfig): {
  is_valid: boolean;
  errors: string[];
} {
  const errors: string[] = [];

  if (config.max_position_per_market <= 0) {
    errors.push('max_position_per_market must be positive');
  }

  if (config.max_total_position <= 0) {
    errors.push('max_total_position must be positive');
  }

  if (config.max_total_position < config.max_position_per_market) {
    errors.push('max_total_position must be >= max_position_per_market');
  }

  if (config.max_daily_loss_usd <= 0) {
    errors.push('max_daily_loss_usd must be positive');
  }

  if (config.max_loss_per_trade_usd < 0) {
    errors.push('max_loss_per_trade_usd must be non-negative');
  }

  if (config.max_consecutive_errors <= 0) {
    errors.push('max_consecutive_errors must be positive');
  }

  if (config.gas_buffer_cents < 0) {
    errors.push('gas_buffer_cents must be non-negative');
  }

  if (config.liquidity_factor <= 0 || config.liquidity_factor > 1) {
    errors.push('liquidity_factor must be in (0, 1]');
  }

  return {
    is_valid: errors.length === 0,
    errors,
  };
}

File: backend/app/engines/circuit_breaker.py
"""
Circuit Breaker and Risk Management System
Migrated from Rust bot circuit_breaker.rs

Implements production-grade risk management with configurable limits and automatic shutoffs

Key Features:
- Position limits per market and total
- Daily loss limits with automatic trading halt
- Per-trade loss limits
- Consecutive error tracking
- Gas cost buffers
- Automatic circuit breaker trip and recovery
"""

import logging
from typing import Dict, Optional, List
from datetime import datetime, date
from dataclasses import dataclass, field

from app.models.arbitrage import (
    CircuitBreakerConfig,
    CircuitBreakerState,
    ValidationResult,
    TradeResult,
    DailyMetrics,
    CircuitBreakerStatus,
    Position,
)

logger = logging.getLogger(__name__)


class CircuitBreaker:
    """
    Circuit breaker for trading risk management

    State transitions:
    - CLOSED: Trading enabled (normal operation)
    - OPEN: Trading disabled (circuit tripped)
    - HALF_OPEN: Testing if conditions improved

    Automatic transitions:
    - CLOSED → OPEN: When limits exceeded
    - OPEN → HALF_OPEN: After cooldown period
    - HALF_OPEN → CLOSED: When conditions improved
    """

    def __init__(self, config: Optional[CircuitBreakerConfig] = None):
        """
        Initialize circuit breaker

        Args:
            config: Circuit breaker configuration (uses defaults if None)
        """
        self.config = config or CircuitBreakerConfig()
        self.state = CircuitBreakerState.CLOSED
        self.positions: Dict[str, Position] = {}
        self.daily_metrics = self._initialize_daily_metrics()
        self.trip_time: Optional[int] = None
        self.error_count = 0
        self.last_reset_time = int(datetime.now().timestamp() * 1000)

        logger.info(f"[CIRCUIT BREAKER] Initialized with state: {self.state.value}")

    # ========================================================================
    # STATE MANAGEMENT
    # ========================================================================

    def get_state(self) -> CircuitBreakerState:
        """
        Get current circuit breaker state with auto-transitions

        Returns:
            Current CircuitBreakerState
        """
        now_ms = int(datetime.now().timestamp() * 1000)

        # Auto-transition from HALF_OPEN to CLOSED
        if self.state == CircuitBreakerState.HALF_OPEN and self._can_reset():
            self.state = CircuitBreakerState.CLOSED
            self.error_count = 0
            self.trip_time = None
            logger.info("[CIRCUIT BREAKER] HALF_OPEN → CLOSED (conditions improved)")

        # Auto-transition from OPEN to HALF_OPEN after cooldown
        if (
            self.state == CircuitBreakerState.OPEN
            and self.trip_time
            and (now_ms - self.trip_time) > self.config.error_cooldown_ms
        ):
            self.state = CircuitBreakerState.HALF_OPEN
            self.error_count = max(1, self.error_count // 2)  # Reduce error count
            logger.info(f"[CIRCUIT BREAKER] OPEN → HALF_OPEN (cooldown elapsed)")

        return self.state

    def can_trade(self) -> bool:
        """
        Check if trading is allowed

        Returns:
            True if trading allowed, False otherwise
        """
        state = self.get_state()
        return state in [CircuitBreakerState.CLOSED, CircuitBreakerState.HALF_OPEN]

    def _can_reset(self) -> bool:
        """Check if circuit breaker can be reset to CLOSED"""
        return (
            self.error_count == 0
            and self.daily_metrics.consecutive_errors < self.config.max_consecutive_errors
            and self.daily_metrics.total_pnl_usd > -self.config.max_daily_loss_usd
        )

    def _trip(self, reason: str) -> None:
        """
        Trip the circuit breaker (halt trading)

        Args:
            reason: Reason for tripping
        """
        self.state = CircuitBreakerState.OPEN
        self.trip_time = int(datetime.now().timestamp() * 1000)
        logger.error(f"[CIRCUIT BREAKER] TRIPPED: {reason}")

    # ========================================================================
    # TRADE VALIDATION
    # ========================================================================

    def validate_trade(
        self,
        market_id: str,
        trade_size_usd: float,
        estimated_loss_usd: float,
    ) -> ValidationResult:
        """
        Validate if a trade can be executed

        Checks:
        1. Circuit breaker state
        2. Position limits (per market and total)
        3. Daily loss limit
        4. Per-trade loss limit

        Args:
            market_id: Market identifier
            trade_size_usd: Trade size in dollars
            estimated_loss_usd: Estimated maximum loss in dollars

        Returns:
            ValidationResult with can_execute flag and reason
        """
        # Check circuit breaker state
        if not self.can_trade():
            return ValidationResult(
                can_execute=False,
                reason=f"Circuit breaker is {self.state.value}",
            )

        # Check daily loss limit
        projected_loss = self.daily_metrics.total_pnl_usd - estimated_loss_usd
        if projected_loss < -self.config.max_daily_loss_usd:
            self._trip(f"Daily loss limit exceeded: ${projected_loss:.2f}")
            return ValidationResult(
                can_execute=False,
                reason="Daily loss limit would be exceeded",
            )

        # Check per-trade loss limit
        if estimated_loss_usd > self.config.max_loss_per_trade_usd:
            return ValidationResult(
                can_execute=False,
                reason=f"Per-trade loss limit exceeded: ${estimated_loss_usd:.2f}",
            )

        # Check position limits
        current_position = self.positions.get(market_id)
        current_position_size = (
            current_position.quantity if current_position else 0
        )
        new_position_size = current_position_size + trade_size_usd

        if new_position_size > self.config.max_position_per_market:
            return ValidationResult(
                can_execute=False,
                reason=f"Position limit for market would be exceeded: "
                f"{new_position_size:.0f} > {self.config.max_position_per_market}",
            )

        # Check total position limit
        total_position = self._calculate_total_position()
        if total_position + trade_size_usd > self.config.max_total_position:
            return ValidationResult(
                can_execute=False,
                reason="Total position limit would be exceeded",
            )

        return ValidationResult(can_execute=True)

    # ========================================================================
    # POSITION TRACKING
    # ========================================================================

    def update_position(self, market_id: str, result: TradeResult) -> None:
        """
        Update position after trade execution

        Args:
            market_id: Market identifier
            result: Trade execution result
        """
        if not result.success:
            self._handle_error(result.error_message or "Unknown error")
            return

        # Reset error count on success
        self.error_count = 0
        self.daily_metrics.consecutive_errors = 0

        # Update position (simplified - in real system would track actual fills)
        if result.position and not self.positions.get(market_id):
            self.positions[market_id] = result.position
            logger.info(f"[POSITION TRACKER] Added position: {result.position.id}")

    def get_position(self, market_id: str) -> Optional[Position]:
        """
        Get position for a market

        Args:
            market_id: Market identifier

        Returns:
            Position if exists, None otherwise
        """
        return self.positions.get(market_id)

    def get_all_positions(self) -> List[Position]:
        """Get all open positions"""
        return list(self.positions.values())

    def _calculate_total_position(self) -> float:
        """Calculate total position size across all markets"""
        total = 0.0
        for position in self.positions.values():
            total += position.quantity
        return total

    def calculate_unrealized_pnl(self) -> float:
        """
        Calculate unrealized P&L for all positions

        Returns:
            Total unrealized P&L in USD
        """
        total_pnl = 0.0
        for position in self.positions.values():
            total_pnl += position.unrealized_pnl_usd
        return total_pnl

    # ========================================================================
    # ERROR HANDLING
    # ========================================================================

    def _handle_error(self, error_message: str) -> None:
        """
        Handle trade execution error

        Args:
            error_message: Error description
        """
        self.error_count += 1
        self.daily_metrics.consecutive_errors += 1
        self.daily_metrics.failed_trades += 1

        logger.error(
                    f"[CIRCUIT BREAKER] Error ({self.error_count}/{self.config.max_consecutive_errors}): {error_message}"
        )

        # Trip circuit breaker if too many consecutive errors
        if self.error_count >= self.config.max_consecutive_errors:
            self._trip(f"Too many consecutive errors: {self.error_count}")

    def record_success(self, trade_result: TradeResult) -> None:
        """
        Record successful trade

        Args:
            trade_result: Trade execution result
        """
        self.daily_metrics.successful_trades += 1
        self.daily_metrics.total_trades += 1

        # Update P&L
        if trade_result.position:
            self.daily_metrics.total_pnl_usd += trade_result.position.unrealized_pnl_usd

        # Track gas costs
        if trade_result.gas_cost_usd:
            self.daily_metrics.total_gas_spent_usd += trade_result.gas_cost_usd

        # Reset error count on success
        self.error_count = 0
        self.daily_metrics.consecutive_errors = 0

        logger.info(f"[CIRCUIT BREAKER] Trade recorded successfully")

    # ========================================================================
    # METRICS AND REPORTING
    # ========================================================================

    def get_daily_metrics(self) -> DailyMetrics:
        """
        Get daily metrics

        Returns:
            Current DailyMetrics
        """
        # Reset if new day
        today = date.today().isoformat()
        if self.daily_metrics.date != today:
            self.daily_metrics = self._initialize_daily_metrics()

        return self.daily_metrics

    def _initialize_daily_metrics(self) -> DailyMetrics:
        """Initialize daily metrics for new day"""
        today = date.today().isoformat()
        logger.info(f"[CIRCUIT BREAKER] Initialized daily metrics for {today}")
        return DailyMetrics(
            date=today,
            total_trades=0,
            successful_trades=0,
            failed_trades=0,
            total_pnl_usd=0.0,
            max_drawdown_usd=0.0,
            consecutive_errors=0,
            total_gas_spent_usd=0.0,
        )

    def get_status(self) -> CircuitBreakerStatus:
        """
        Get circuit breaker status

        Returns:
            CircuitBreakerStatus with current state
        """
        metrics = self.get_daily_metrics()
        state = self.get_state()

        return CircuitBreakerStatus(
            state=state,
            can_trade=self.can_trade(),
            error_count=self.error_count,
            consecutive_errors=metrics.consecutive_errors,
            daily_pnl_usd=metrics.total_pnl_usd,
            daily_loss_remaining_usd=self.config.max_daily_loss_usd
            - abs(metrics.total_pnl_usd),
            total_positions=len(self.positions),
            trip_time=self.trip_time,
        )

    def get_diagnostics(self) -> dict:
        """
        Get detailed diagnostics

        Returns:
            Dictionary with diagnostic information
        """
        return {
            "config": {
                "max_position_per_market": self.config.max_position_per_market,
                "max_total_position": self.config.max_total_position,
                "max_daily_loss_usd": self.config.max_daily_loss_usd,
                "max_loss_per_trade_usd": self.config.max_loss_per_trade_usd,
                "max_consecutive_errors": self.config.max_consecutive_errors,
                "error_cooldown_ms": self.config.error_cooldown_ms,
                "gas_buffer_cents": self.config.gas_buffer_cents,
                "liquidity_factor": self.config.liquidity_factor,
            },
            "state": self.get_state().value,
            "positions": [p.__dict__ for p in self.get_all_positions()],
            "daily_metrics": {
                "date": self.daily_metrics.date,
                "total_trades": self.daily_metrics.total_trades,
                "successful_trades": self.daily_metrics.successful_trades,
                "failed_trades": self.daily_metrics.failed_trades,
                "total_pnl_usd": self.daily_metrics.total_pnl_usd,
                "consecutive_errors": self.daily_metrics.consecutive_errors,
                "total_gas_spent_usd": self.daily_metrics.total_gas_spent_usd,
            },
            "status": self.get_status().__dict__,
        }

    # ========================================================================
    # MANUAL CONTROLS
    # ========================================================================

    def manual_reset(self) -> None:
        """Manually reset circuit breaker (use with caution)"""
        logger.warning("[CIRCUIT BREAKER] Manual reset requested")
        self.state = CircuitBreakerState.CLOSED
        self.error_count = 0
        self.trip_time = None
        self.daily_metrics.consecutive_errors = 0
        logger.info("[CIRCUIT BREAKER] Manually reset to CLOSED")

    def manual_trip(self, reason: str = "Manual trip") -> None:
        """Manually trip circuit breaker (emergency stop)"""
        logger.warning(f"[CIRCUIT BREAKER] Manual trip requested: {reason}")
        self._trip(f"Manual: {reason}")

    def update_config(self, **kwargs) -> None:
        """
        Update configuration parameters

        Args:
            **kwargs: Configuration parameters to update
        """
        for key, value in kwargs.items():
            if hasattr(self.config, key):
                setattr(self.config, key, value)
                logger.info(f"[CIRCUIT BREAKER] Updated config: {key} = {value}")
            else:
                logger.warning(f"[CIRCUIT BREAKER] Unknown config key: {key}")


# ============================================================================
# UTILITY FUNCTIONS
# ============================================================================

def create_circuit_breaker(**kwargs) -> CircuitBreaker:
    """
    Create a circuit breaker instance

    Args:
        **kwargs: Configuration parameters

    Returns:
        CircuitBreaker instance
    """
    config = CircuitBreakerConfig(**kwargs)
    return CircuitBreaker(config)


def validate_config(config: CircuitBreakerConfig) -> dict:
    """
    Validate circuit breaker configuration

    Args:
        config: Configuration to validate

    Returns:
        Dictionary with is_valid flag and errors list
    """
    errors = []

    if config.max_position_per_market <= 0:
        errors.append("max_position_per_market must be positive")

    if config.max_total_position <= 0:
        errors.append("max_total_position must be positive")

    if config.max_total_position < config.max_position_per_market:
        errors.append("max_total_position must be >= max_position_per_market")

    if config.max_daily_loss_usd <= 0:
        errors.append("max_daily_loss_usd must be positive")

    if config.max_loss_per_trade_usd < 0:
        errors.append("max_loss_per_trade_usd must be non-negative")

    if config.max_consecutive_errors <= 0:
        errors.append("max_consecutive_errors must be positive")

    if config.gas_buffer_cents < 0:
        errors.append("gas_buffer_cents must be non-negative")

    if not (0 < config.liquidity_factor <= 1):
        errors.append("liquidity_factor must be in (0, 1]")

    return {
        "is_valid": len(errors) == 0,
        "errors": errors,
    }
